fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +716,193 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+    public static PredicateSplit splitPredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return PredicateSplit.of(null, null, null);
+        }
+        List<ScalarOperator> predicateConjuncts = Utils.extractConjuncts(predicate);
+        List<ScalarOperator> columnEqualityPredicates = Lists.newArrayList();
+        List<ScalarOperator> rangePredicates = Lists.newArrayList();
+        List<ScalarOperator> residualPredicates = Lists.newArrayList();
+        for (ScalarOperator scalarOperator : predicateConjuncts) {
+            if (scalarOperator instanceof BinaryPredicateOperator) {
+                BinaryPredicateOperator binary = (BinaryPredicateOperator) scalarOperator;
+                ScalarOperator leftChild = scalarOperator.getChild(0);
+                ScalarOperator rightChild = scalarOperator.getChild(1);
+                if (binary.getBinaryType().isEqual()) {
+                    if (leftChild.isColumnRef() && rightChild.isColumnRef()) {
+                        columnEqualityPredicates.add(scalarOperator);
+                    } else if (leftChild.isColumnRef() && rightChild.isConstantRef()) {
+                        rangePredicates.add(scalarOperator);
+                    } else {
+                        residualPredicates.add(scalarOperator);
+                    }
+                } else if (binary.getBinaryType().isRange()) {
+                    if (leftChild.isColumnRef() && rightChild.isConstantRef()) {
+                        rangePredicates.add(scalarOperator);
+                    } else {
+                        residualPredicates.add(scalarOperator);
+                    }
+                }
+            } else {
+                residualPredicates.add(scalarOperator);
+            }
+        }
+        return PredicateSplit.of(Utils.compoundAnd(columnEqualityPredicates), Utils.compoundAnd(rangePredicates),
+                Utils.compoundAnd(residualPredicates));
+    }
+
+    public static ScalarOperator canonizePredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
+        ScalarOperatorRewriter rewrite = new ScalarOperatorRewriter();
+        return rewrite.rewrite(predicate, ScalarOperatorRewriter.DEFAULT_REWRITE_SCAN_PREDICATE_RULES);
+    }
+
+    public static ScalarOperator canonizePredicateForRewrite(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
+        ScalarOperatorRewriter rewrite = new ScalarOperatorRewriter();
+        return rewrite.rewrite(predicate, ScalarOperatorRewriter.MV_SCALAR_REWRITE_RULES);
+    }
+
+    public static boolean isAlwaysFalse(ScalarOperator predicate) {
+        ScalarOperatorRewriter rewriter = new ScalarOperatorRewriter();
+        return ConstantOperator.FALSE.equals(rewriter.rewrite(predicate,
+                ScalarOperatorRewriter.DEFAULT_REWRITE_SCAN_PREDICATE_RULES));
+    }
```

Comment:
BTW，I think your code only need check is ConstantOperator.TRUE.equals(predicate), don't need handle compound because you had rewrite

Reply:
no, the rewritten predicates can also have compound. eg: true and true. So here should handle compound.

Reply:
I found your usage in MVRewirteRule, there can never be `true and true`

Reply:
ok. I will remove it.